Add database injection

keiranjprice101 opened the pull request:
### Summary of work
Majority of commits add a way to inject data before the action runs. An explanation of this is within the README commit 8755ac4

 - 6b092e2 This commit relates to issue #1036
 - 5fc3f06  is also related to #1036 as most data will be coming from the database now
 - 1adc97f This commit turns off duplication search within pylint due to false positives.




### How to test your work
<!---This can be a link to the---> 

### Additional comments
<!---Anything else: e.g. was the estimate reasonable for this issue?---> 

closes #986


**Before merging ensure the release notes have been updated**

dtasev commented on webtests/database.py:
```diff
@@ -0,0 +1,106 @@
+    :returns: (Connection) The PyMySql Connection object.
+
+    """
+    return pymysql.connect(host=get_str("DATABASE", "host"),
+                           user=get_str("DATABASE", "user"),
+                           password=get_str("DATABASE", "password"),
+                           database=get_str("DATABASE", "name"),
+                           charset='utf8mb4',
+                           cursorclass=pymysql.cursors.DictCursor)
```
Can this use Django's ORM?

dtasev commented on webtests/database.py:
```diff
@@ -0,0 +1,106 @@
+    :param dataset: (Path) Path object for dataset
+    """
+    dataset = parse_dataset(dataset)
+    queries = ["INSERT INTO %s (%s) VALUES (%s)" %
```
Can we avoid any raw SQL and go through the ORM?

dtasev reviewed (changes requested):
Can we not get this through the Django ORM? The injection here seems to be using pymysql to directly insert into the DB, but that won't work for anything but MySQL. 

With https://github.com/ISISScientificComputing/autoreduce/issues/1072 I'm removing use of SQLAlchemy, and then we'll be using the Django ORM everywhere, which makes it super easy to switch backends (e.g. to SQLite3 for fast testing & local running).

keiranjprice101 commented:
@dtasev 
Tests now use `StaticLiveServerTestCase` which self deploys. Data now comes from fixtures. Also changed `url` to `re_path` as `url` is deprecated.

Updated README to reflect to instructions. 

Best to check files changed not commits. I had a bit of a dull moment with a merge conflict.

dtasev commented:
You'll have to rebase again - a lot of these urls don't exist anymore, they are now included from `instrument.urls` (all `instrument/`) and `reduction_viewer.urls` (most `runs/`)

dtasev commented on WebApp/autoreduce_webapp/selenium_tests/pages/instrument_summary_page.py:
```diff
@@ -0,0 +1,68 @@
+            if run.text == run_string:
+                run.click()
+                return RunSummaryPage(self.driver, self.instrument, run_number, version)
+        raise NoSuchElementException
```
That seems like a bad merge - will fix

keiranjprice101 commented on WebApp/autoreduce_webapp/autoreduce_webapp/settings.py:
```diff
@@ -105,24 +106,27 @@ def get_str(section, key):
-    'default': {
-        'ENGINE': 'django.db.backends.sqlite3',  # Add 'postgresql_psycopg2', 'mysql', 'sqlite3' or 'oracle'.
-        'NAME': f'{PROJECT_ROOT}/sqlite3.db',  # Or path to database file if using sqlite3.
+if 'test' in sys.argv:
```
Needs changing

keiranjprice101 commented on .github/workflows/run-tests.yml:
```diff
@@ -46,6 +46,19 @@ jobs:
   selenium:
     name: Selenium
     runs-on: ubuntu-18.04
+    services:
```
Needs changing

dtasev commented on WebApp/autoreduce_webapp/selenium_tests/configuration.py:
```diff
@@ -15,8 +15,8 @@
 
-CONFIG_PATH = Path(get_project_root(), "webtests/config.json")
-TEMP_CONFIG_PATH = Path(get_project_root(), "webtests/temp_config.json")
+CONFIG_PATH = Path(get_project_root(), "WebApp/autoreduce_webapp/selenium_tests/config.json")
+TEMP_CONFIG_PATH = Path(get_project_root(), "Webapp/autoreduce_webapp/selenium_tests/temp_config.json")
```
Should use `PROJECT_ROOT` with `from utils.project.structure import PROJECT_ROOT`

keiranjprice101 replied:
Aye, this wasn't available originally. Will add it Cheers 👍

dtasev commented:
The fail to find All instrument, failed jobs etc (from https://github.com/ISISScientificComputing/autoreduce/pull/1044/checks?check_run_id=1917570724) makes me think I might have messed up something with the login in the tests - you don't see those buttons unless logged in

dtasev commented:
Can we remove the requirement to have to manually create a config.json? It seems like it gets overwritten for every test anyway so it doesn't seem necessary anymore

dtasev reviewed (changes requested):
~~The latest changes fixed running headless for me, but if I switch headless to false I get:~~

```
selenium.common.exceptions.WebDriverException: Message: unknown error: Chrome failed to start: exited abnormally.
  (unknown error: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /usr/bin/google-chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
```

~~I think it'd be useful having the option to not run headless, especially during writing tests, but also for debugging. Any way we can have both?~~

dtasev commented:
Nevermind that one's on me - I was running it from SSH but hadn't set the DISPLAY env variable. Actually that might be a good check to have in `get_chrome_driver`!

dtasev reviewed (approved):
Looks good, runs fine locally both headless and not.
